File: .claude/skills/media-use/audio/scripts/lib/gemini-auth.py
"""Obtain Gemini OAuth headers without persisting credentials or tokens."""

import json
import os
import sys

# Keep the credential subprocess protocol UTF-8 on every platform.
for _stream in (sys.stdout, sys.stderr):
    if hasattr(_stream, "reconfigure"):
        _stream.reconfigure(encoding="utf-8", errors=_stream.errors)


def authenticate():
    try:
        from google.auth.transport.requests import Request
        from google.oauth2 import service_account
    except ImportError:
        return {"error": "Service-account auth needs Python packages: pip install google-auth requests"}

    try:
        path = os.environ.get("GOOGLE_APPLICATION_CREDENTIALS")
        if path:
            with open(path, encoding="utf-8") as source:
                info = json.load(source)
        else:
            info = json.loads(os.environ["GCS_CREDS"])
        if info.get("type") != "service_account":
            return {"error": "Gemini requires service-account JSON, not user ADC credentials"}
        # Only exchange credentials with Google's OAuth server.
        info["token_uri"] = "https://oauth2.googleapis.com/token"
        credentials = service_account.Credentials.from_service_account_info(
            info,
            scopes=[
                "https://www.googleapis.com/auth/generative-language.retriever",
            ],
        )
        credentials.refresh(Request())
        project = (
            os.environ.get("GOOGLE_CLOUD_PROJECT")
            or os.environ.get("GCLOUD_PROJECT_ID")
            or info.get("project_id")
        )
        return {"token": credentials.token, "project": project}
    except Exception:
        # Library exceptions can contain credential data. Never forward them.
        return {"error": "Gemini service-account authentication failed; check credentials, permissions, and network access"}


if __name__ == "__main__":
    result = authenticate()
    print(json.dumps(result))
    sys.exit(1 if "error" in result else 0)

File: .claude/skills/media-use/audio/scripts/lib/gemini-auth.mjs
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";
import { pythonInvocation } from "./python.mjs";

export function geminiConfigured() {
  return Boolean(
    process.env.GEMINI_API_KEY ||
    process.env.GOOGLE_API_KEY ||
    process.env.GOOGLE_APPLICATION_CREDENTIALS ||
    process.env.GCS_CREDS,
  );
}

export function geminiAuth({ run = spawnSync } = {}) {
  const key = process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY;
  if (key) return { headers: { "x-goog-api-key": key }, secret: key };
  if (!geminiConfigured()) {
    throw new Error(
      "Gemini TTS needs GEMINI_API_KEY or GOOGLE_API_KEY, or service-account credentials (GOOGLE_APPLICATION_CREDENTIALS or GCS_CREDS)",
    );
  }
  const { cmd, args } = pythonInvocation([
    fileURLToPath(new URL("./gemini-auth.py", import.meta.url)),
  ]);
  const result = run(cmd, args, { encoding: "utf8", timeout: 60_000, maxBuffer: 1024 * 1024 });
  let data;
  try {
    data = JSON.parse(result.stdout);
  } catch {
    throw new Error(
      "Gemini service-account authentication failed; check Python and install google-auth requests",
    );
  }
  if (result.status !== 0 || !data.token) {
    // Do not echo subprocess output: it could contain credentials.
    throw new Error(
      "Gemini service-account authentication failed; check service-account JSON, permissions, and Python packages google-auth requests",
    );
  }
  const headers = { Authorization: `Bearer ${data.token}` };
  if (data.project) headers["x-goog-user-project"] = data.project;
  return { headers, secret: data.token };
}
